src/rgw/rgw_auth_keystone.cc

```diff
@@ -167,8 +166,14 @@ TokenEngine::get_acl_strategy(const TokenEngine::token_envelope_t& token) const
     make_spec_item("*", user_name),
   };
 
+  std::list<rgw::keystone::TokenEnvelope::Role> token_roles;
+  for (const auto& r : token.roles) {
+    token_roles.insert(token_roles.end(),
+        rgw::keystone::TokenEnvelope::Role(r));
+  }
+
   /* Lambda will obtain a copy of (not a reference to!) allowed_items. */
-  return [allowed_items](const rgw::auth::Identity::aclspec_t& aclspec) {
+  return [allowed_items, token_roles](const rgw::auth::Identity::aclspec_t& aclspec) {
```

Comment:
Isn't it possible to re-use token.roles in the lambda, instead of constructing a new 'token_roles' list?

Reply:
Not being a C++ programmer, I don't know what rules apply exactly to what is captured for a lambda. Roles as such as an std::list in the TokenEnvelope. If they were a reference to begin with, then sure. I could stand an expert help.

Reply:
@zaitcev you can look at a C++ lambda as a struct combined with a function (where the members of the struct are the captured values). so, you should not capture a reference or pointer to a local variable in a function and use the lambda outside of the function. but you can capture a reference to something that is passed by reference into the function.
the caller of the function will have to use the lambda before the parameter's (the `token` in this case) lifetime ends, but this can be part of the contract of this function.